### ISSUE TITLE
js.html.MediaElement has Incorrect Interface

### ISSUE THREAD

**Issue by alexmdodge** [HaxeFoundation/haxe#6787]
## Overview
`js.html.VideoElement` which extends `js.html.MediaElement` has an incorrect return type for the `play()` API. Current documentation states that the interface should be defined as follows,

```js
function play() : Promise<Void>;
```

This is supported by the respective `mozilla\HTMLMediaElement.webidl`, as well as the MDN documentation, https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/play

## Solution
It seems it has been some time since these externs have been re-generated using the `html-externs` repository.

**Comment by SlavaRa** [comment]
Not for all browsers...

**Comment by alexmdodge** [comment]
Very true, https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/play#Browser_compatibility, but is it possible to have a custom union type? Or provide a way of extending the interface such that a returned promise is supported?

It also wouldn't be detrimental as those working in older browsers would not use the play functionality with a return type, and checks could be made client side to ascertain whether it is viable.

As some context, I am working on a project which requires autoplay support for video elements.

**Comment by SlavaRa** [comment]
Maybe it will help you
```haxe
var promise = untyped __js__('
	this.audioElement.play();
');
if(promise != null) {
	untyped __js__('
		promise.catch(...);
	');
}
```

**Comment by alexmdodge** [comment]
@SlavaRa Thank you for that snippet! I had considered that, and it most certainly works. I instead opted to `extend` the html `Element` extern class and redefine the `MediaElement` entirely, just to enforce the type checking we have throughout project modules. We have a lot of `MediaElement` interactions throughout our codebase and having a strict contract is preferred to untyped sections.

**Comment by nadako** [comment]
why do `untyped __js__`, even twice?

```haxe
var p:Promise<Void> = untyped e.play();
p.then(_ -> trace("done"));
```

and then you can use static extensions so you don't go untyped each time: https://try.haxe.org/#b3cf5

---

anyway, maybe we should just change the return type to `Promise<Void>`

**Comment by SlavaRa** [comment]
mb Null<Promise<Void>>?
```haxe
var p:Promise<Void> = untyped e.play(); // null for IE, Edge...
```

**Comment by alexmdodge** [comment]
@nadako I appreciate the second code example! I'm most certainly not blocked anymore, so thank you both for your help. I would say that in general it would be good to update the interfaces. If they are based on the Mozilla defined `webidl` files then they should reflect those definitions (aside from conflicting implementations and unnecessary definitions defined in the `html-externs` repo [readme](https://github.com/HaxeFoundation/html-externs)).

Thanks once again!

**Comment by haxiomic** [comment]
Resolved in #7354
